Add Faking HttpClient recipe

blairconrad opened the pull request:
Adds runnable recipes (in form of tests) and includes those in a Recipe section of the documentation, by reading "snippets" right out of the executable recipe.

----
**Old description:**
Not baked yet, but a good starting point for discussion. In particular, we're going to at least want to decide whether we want these things called "case studies" or "recipes" or whatever. I picked one so I could have something to commit and push.
@thomaslevesque, everything's negotiable. For example, you might hate tone, approach of showing one way to work and then introducing the fakeable context, showing a "quicky" version of the test where we exercise the fake directly, anything. Don't hold back.

I also have ideas for hosting the code in this article (or others?) separately, similarly to the samples directory, but probably runnable. But let's look at that later…

thomaslevesque commented on docs/Recipes/faking-http-client.md:
```diff
@@ -0,0 +1,98 @@
+
+Let's assume that you want to create a fake `HttpClient` so you can dictate the
+behavior of the
+[GetAsync(String)](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.getasync?view=net-7.0#system-net-http-httpclient-getasync(system-string))
```
I realize the goal isn't to cover every scenario, but it might be helpful to also include an example for `PostAsync` with a request body.

blairconrad replied:
I'll work something up.

blairconrad replied:
Is this close to what you were hoping for?

```csharp
public async Task PostAsync()
{
    using var response = new HttpResponseMessage
    {
        Content = new StringContent("FakeItEasy is fun")
    };

    var handler = A.Fake<FakeableHttpMessageHandler>();

    using var client = new HttpClient(handler);

    using var postContent = new StringContent("I am posting this.");
    var result = await client.PostAsync("https://fakeiteasy.github.io/docs/", postContent);

    A.CallTo(() => handler.FakeSendAsync(
            A<HttpRequestMessage>.That.Matches(
                m => m.Content!.ReadAsStringAsync().Result == "I am posting this."),
            A<CancellationToken>.Ignored))
        .MustHaveHappened();
}
```

thomaslevesque replied:
Looks good to me! The "sync over async" bit is a little annoying, but I don't think there's much we can do about it... I don't think it can actually cause a deadlock in this case.

thomaslevesque replied:
yes

thomaslevesque commented on docs/Recipes/faking-http-client.md:
````diff
@@ -0,0 +1,98 @@
+
+```csharp
+public Task<HttpResponseMessage>
+    GetAsync([StringSyntax(StringSyntaxAttribute.Uri)] string? requestUri)
````
I think we can omit the attribute, for the sake of readability.

thomaslevesque commented on docs/Recipes/faking-http-client.md:
```diff
@@ -0,0 +1,98 @@
+
+With this knowledge, we can write a passing test:
+
+??? note "This is a simplified example"
```
Does `??? node` do something special that I'm not aware of? Or is it just a placeholder for something that will actually render as a note?

blairconrad replied:
It does something you're not aware of! I thought you'd like it.

![image](https://user-images.githubusercontent.com/3275797/225489817-ccf121de-64ab-4b17-badb-cd4d5a616baf.png)

![image](https://user-images.githubusercontent.com/3275797/225489850-2145d38b-0947-4bcb-8f6d-a95948713c21.png)

thomaslevesque replied:
I do like it!
It's specific to mkdocs, right?

blairconrad replied:
Specifically it's a [PyMdown Extention](https://facelessuser.github.io/pymdown-extensions/extensions/details/).

thomaslevesque commented on docs/Recipes/faking-http-client.md:
````diff
@@ -0,0 +1,98 @@
+```csharp
+public async Task Test()
+{
+    var response = new HttpResponseMessage();
+    response.Content = new StringContent("FakeItEasy is fun");
````
nit: maybe use object initializer syntax here?
```suggestion
    var response = new HttpResponseMessage
    {
        Content = new StringContent("FakeItEasy is fun")
    };

```

blairconrad replied:
I originally had it, but one one line, and it didn't render well. Will try your version.

thomaslevesque reviewed (approved):
Thanks! This looks great. I have a few suggestions, nothing major.

thomaslevesque commented on docs/Recipes/faking-http-client.md:
````diff
@@ -12,7 +12,7 @@ First off, let's look at the declaration of `GetAsync`:
 ```csharp
 public Task<HttpResponseMessage>
-    GetAsync([StringSyntax(StringSyntaxAttribute.Uri)] string? requestUri)
+    GetAsync(string? requestUri)
````
Without the attribute, you could put it all on one line

blairconrad commented:
Hey, @thomaslevesque. Still a WIP, and builds on the yet-unmerged mkdocs upgrade.
But I thought you might enjoy seeing the slightly different route I took. In particular, note that the recipes are working tests, and I include snippets in the docs. I think this is cool. Also maybe useful.

Oh, the recipe sometimes fail on net4.5. Possibly due to sync over async?
I haven't gotten far into debugging. Part of me wonders whether we should even worry about having the recipes work on all TFMs. Another part of me says "well, yeah, if we can". Another part worries that the net45 failure may rear its ugly head on other TFMs even though I haven't seen it yet locally.

blairconrad commented on recipes/FakeItEasy.Recipes.CSharp/FakingHttpClient.cs:
```diff
@@ -48,24 +51,29 @@ public async Task FakeByMakingMessageHandlerFakeable()
 
-        using var postContent = new StringContent("I am posting this.");
-        var result = await client.PostAsync("https://fakeiteasy.github.io/docs/", postContent);
+        using var postContent = new StringContent("my post");
```
the original was a little wide when rendered

blairconrad commented:
As usual, I made a few assumptions that I'm happy to have overturned. Maybe you don't like the recipes being tests. Maybe you want a different folder organization. I dunno. Whatever.
Also, I'd probably arrange the commits differently, for a final merge, just going right to the snippetted version (if you like the snippets), but I thought that at least this way I can drop commits if you hate something.

thomaslevesque commented on recipes/FakeItEasy.Recipes.CSharp/FakeItEasy.Recipes.CSharp.csproj:
Do we actually need a new project for this? Maybe we could add the recipes to the examples project? (I'm not really opposed to a new project, just thinking out loud)

thomaslevesque replied:
> "Hey! The examples were useful back when we had no docs, and are pretty thin now. What if we just made sure everything was documented or a recipe, if deserving, and remove that project"

I kinda like this idea. I doubt many people (if any) actually look at the Examples project anyway...

blairconrad replied:
Thanks. I never do! Created #1917.

thomaslevesque reviewed (commented):
> In particular, note that the recipes are working tests, and I include snippets in the docs. I think this is cool.

It _is_ pretty cool! Nice work

> Oh, the recipe sometimes fail on net4.5. Possibly due to sync over async?

This is annoying... I'll try to see what's going on.

thomaslevesque commented on recipes/FakeItEasy.Recipes.CSharp/FakingHttpClient.cs:
```diff
@@ -0,0 +1,88 @@
+        using var client = new HttpClient(handler);
+
+        using var postContent = new StringContent("my post");
+        var result = await client.PostAsync(
```
nit: the `result` variable is never used

blairconrad commented:
> > Oh, the recipe sometimes fail on net4.5. Possibly due to sync over async?
> 
> This is annoying... I'll try to see what's going on.

Copying in your comments from sidechat. 

----

Looks like in .NET 4.5, the content is disposed as soon as the request is sent :/
https://github.com/dotnet/runtime/blob/b0b7aaefb88aa8d01b3d64fb40ac2f73a9d98c3e/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs#L665-L680
I think the best path forward is to warn that it doesn't work in .NET 4.5 (and ensure it actually work for all the other TFMs), and skip this test when the TFM is net45
Another option would be to introduce a wrapper that prevent disposal of the content:


```csharp

using var postContent = new StringContent("my post");
var result = await client.PostAsync(
    "https://fakeiteasy.github.io/docs/",
    postContent.PreventDisposal());
```

where `PreventDisposal` is an extension method returning a decorator that overrides `Dispose(bool)` to do nothing

It's not great, though...

More on this: https://github.com/dotnet/runtime/issues/14612


And here's the PR that changes the behavior: https://github.com/dotnet/corefx/pull/19082


I can't figure out in which version it was introduced, though...
Maybe .NET Core 2 (the first preview was released shortly after this PR was merged)

blairconrad commented:
Thanks for the comments and research, @thomaslevesque. I think I've incorporated everything and have squashed and removed the WIP tag.
Of course, still happy to modify.

blairconrad commented on mkdocs.yml:
```diff
@@ -41,8 +41,11 @@ plugins:
-  - pymdownx.snippets
+  - pymdownx.snippets:
+      check_paths: true
+      dedent_subsections: true
```
This is the option that the new material for mkdocs provided.

thomaslevesque commented:
Great job! Thanks @blairconrad
